features/FEATURE_LWIP/lwip-interface/lwip_stack.c

```diff
@@ -369,34 +373,88 @@ static sys_sem_t lwip_netif_has_any_addr;
 #if PREF_ADDR_TIMEOUT
 #define HAS_PREF_ADDR 2
 static sys_sem_t lwip_netif_has_pref_addr;
+#else
+#define HAS_PREF_ADDR 0
 #endif
 #if BOTH_ADDR_TIMEOUT
 #define HAS_BOTH_ADDR 4
 static sys_sem_t lwip_netif_has_both_addr;
+#else
+#define HAS_BOTH_ADDR 0
 #endif
 
-static void mbed_lwip_netif_status_irq(struct netif *lwip_netif)
+static const char lwip_requested_addr_state = HAS_ANY_ADDR | HAS_PREF_ADDR | HAS_BOTH_ADDR;
+
+void mbed_lwip_netif_status_irq(struct netif *lwip_netif)
 {
+    nsapi_connection_status_t status = NSAPI_STATUS_DISCONNECTED;
+    
     if (netif_is_up(lwip_netif)) {
+        bool dns_addr_has_to_be_added = false;
         if (!(lwip_has_addr_state & HAS_ANY_ADDR) && mbed_lwip_get_ip_addr(true, lwip_netif)) {
-            sys_sem_signal(&lwip_netif_has_any_addr);
+            if (lwip_netif->blocking) {
+                sys_sem_signal(&lwip_netif_has_any_addr);
+            }
             lwip_has_addr_state |= HAS_ANY_ADDR;
+            dns_addr_has_to_be_added = true;
         }
 #if PREF_ADDR_TIMEOUT
         if (!(lwip_has_addr_state & HAS_PREF_ADDR) && mbed_lwip_get_ip_addr(false, lwip_netif)) {
-            sys_sem_signal(&lwip_netif_has_pref_addr);
+            if (lwip_netif->blocking) {
+                sys_sem_signal(&lwip_netif_has_pref_addr);
+            }
             lwip_has_addr_state |= HAS_PREF_ADDR;
+            dns_addr_has_to_be_added = true;
         }
 #endif
 #if BOTH_ADDR_TIMEOUT
         if (!(lwip_has_addr_state & HAS_BOTH_ADDR) && mbed_lwip_get_ipv4_addr(lwip_netif) && mbed_lwip_get_ipv6_addr(lwip_netif)) {
-            sys_sem_signal(&lwip_netif_has_both_addr);
+            if (lwip_netif->blocking) {
+                sys_sem_signal(&lwip_netif_has_both_addr);
+            }
             lwip_has_addr_state |= HAS_BOTH_ADDR;
+            dns_addr_has_to_be_added = true;
         }
 #endif
+
+        if (lwip_has_addr_state == lwip_requested_addr_state) {
```

Comment:
In case dual stack, there should be also some timer, that triggers the global address indication/dns server address add, in case only one address is available.